File: dashboard/app/graphql/announcement.ts
export const ANNOUNCEMENT_QUERIES = {
	GET_ANNOUNCEMENTS: `
    query GetAnnouncements($schoolId: String!, $targetType: AnnouncementTargetType, $targetId: String) {
      getAnnouncements(schoolId: $schoolId, targetType: $targetType, targetId: $targetId) {
        id
        schoolId
        authorId
        authorName
        title
        content
        targetType
        targetIds
        priority
        expiresAt
        createdAt
        updatedAt
      }
    }
  `,
	GET_ANNOUNCEMENT: `
    query GetAnnouncement($id: String!) {
      getAnnouncement(id: $id) {
        id
        schoolId
        authorId
        authorName
        title
        content
        targetType
        targetIds
        priority
        expiresAt
        createdAt
        updatedAt
      }
    }
  `,
};

export const ANNOUNCEMENT_MUTATIONS = {
	CREATE_ANNOUNCEMENT: `
    mutation CreateAnnouncement($title: String!, $content: String!, $targetType: AnnouncementTargetType!, $targetIds: [String!]!, $priority: AnnouncementPriority!) {
      createAnnouncement(title: $title, content: $content, targetType: $targetType, targetIds: $targetIds, priority: $priority) {
        id
        title
        content
        createdAt
      }
    }
  `,
	DELETE_ANNOUNCEMENT: `
    mutation DeleteAnnouncement($id: String!) {
      deleteAnnouncement(id: $id)
    }
  `,
};

File: server/src/models/announcement.rs
use async_graphql::*;
use mongodb::bson::oid::ObjectId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Enum)]
pub enum AnnouncementTargetType {
    School,
    Grade,
    Class,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Enum)]
pub enum AnnouncementPriority {
    Low,
    Normal,
    High,
    Urgent,
}

#[derive(Debug, Clone, Serialize, Deserialize, SimpleObject)]
#[graphql(complex)]
pub struct Announcement {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    #[graphql(skip)]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub author_id: String,
    pub author_name: String,
    pub title: String,
    pub content: String,
    pub target_type: AnnouncementTargetType,
    pub target_ids: Vec<String>,
    pub priority: AnnouncementPriority,
    pub expires_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[ComplexObject]
impl Announcement {
    async fn id(&self) -> String {
        self.id.map(|id| id.to_hex()).unwrap_or_default()
    }
}

impl Announcement {
    pub fn new(
        school_id: String,
        author_id: String,
        author_name: String,
        title: String,
        content: String,
        target_type: AnnouncementTargetType,
        target_ids: Vec<String>,
        priority: AnnouncementPriority,
    ) -> Self {
        let now = chrono::Utc::now().to_rfc3339();
        Self {
            id: None,
            school_id,
            author_id,
            author_name,
            title,
            content,
            target_type,
            target_ids,
            priority,
            expires_at: None,
            created_at: now.clone(),
            updated_at: now,
        }
    }
}
